### PR TITLE
Incentives module

### PR DESCRIPTION
**Pull request by teddyknox**

### CODE REVIEW

**Review by jcompagni10 — DISMISSED** [review]

File: proto/duality/incentives/gauge.proto [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  int64 endTick = 3;
+
+  // // Timestamp is used by locks started before the specified duration.
+  // // Timestamp field must not be nil when the lock query type is `ByLockTime`.
+  // // Querying locks with timestamp is currently not implemented.
+  // google.protobuf.Duration stake_duration = 4 [
+  //   (gogoproto.stdtime) = true,
+  //   (gogoproto.nullable) = false,
+  //   (gogoproto.moretags) = "yaml:\"timestamp\""
+  // ];
```
[the remark this answers is not in the document]
Comment (teddyknox) [reply proto/duality/incentives/gauge.proto:82]: Deleting

File: proto/duality/incentives/query.proto [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+option go_package = "github.com/duality-labs/duality/x/incentives/types";
+
+// Query defines the gRPC querier service
+service Query {
+  // GetModuleStatus returns a rundown of coins in the module and their status
+  rpc GetModuleStatus(GetModuleStatusRequest)
+      returns (GetModuleStatusResponse) {
+    option (google.api.http).get =
+        "/duality/incentives/v1beta1/get_module_status";
```
[the remark this answers is not in the document]
Comment (teddyknox) [reply proto/duality/incentives/query.proto:22]: Changed.

File: x/incentives/types/msgs.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+func (m MsgStake) Route() string { return RouterKey }
+func (m MsgStake) Type() string  { return TypeMsgStakeTokens }
+func (m MsgStake) ValidateBasic() error {
+	_, err := sdk.AccAddressFromBech32(m.Owner)
+	if err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "Invalid owner address (%s)", err)
+	}
+
+	// we only allow stakes with one denom for now
+	if m.Coins.Len() != 1 {
```
[the remark this answers is not in the document]
Comment (teddyknox) [reply x/incentives/types/msgs.go:131]: Removing that condition.

File: x/incentives/types/msgs.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "Invalid owner address (%s)", err)
+	}
+
+	for _, unstake := range m.Unstakes {
+		if unstake.ID == 0 {
+			return fmt.Errorf("invalid stakeup ID, got %v", unstake.ID)
+		}
+
+		// only allow unstakes with a single denom or empty
+		if unstake.Coins.Len() > 1 {
```
[the remark this answers is not in the document]
Comment (teddyknox) [reply x/incentives/types/msgs.go:182]: Removing this validation condition as well.

File: x/dex/keeper/pool.go [hunk 5]
```diff
@@ -167,6 +169,7 @@ func (p *Pool) Deposit(
 		// but added error handling to all of the new functions for autoswap.
 		// Open to changing it however.
 		residualShares, _ := p.CalcResidualSharesMinted(residualAmount0, residualAmount1)
+		// TODO: Fix
```
[the remark this answers is not in the document]
Comment (teddyknox) [reply x/dex/keeper/pool.go:172]: Yes we should handle errors generally. Not going to fix in this PR.

File: x/incentives/keeper/hooks.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// BeforeEpochStart is the epoch start hook.
+func (k Keeper) BeforeEpochStart(ctx sdk.Context, epochIdentifier string, epochNumber int64) error {
+	return nil
+}
+
+// AfterEpochEnd is the epoch end hook.
+func (k Keeper) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, epochNumber int64) error {
```
[the remark this answers is not in the document]
Comment (teddyknox) [reply x/incentives/keeper/hooks.go:16]: Ok I added a module parameter limiting the number of gauges. Setting it to 20 for now.

**Review by jcompagni10 — APPROVED** [review]

**Pull request merged**